### PR TITLE
Added dynamic railroad track terrain to Tiberian Sun

### PR DESCRIPTION
**Pull request by Mailaender**
Another small step towards https://github.com/OpenRA/OpenRA/issues/2652.

### CODE REVIEW

**Comment by pchote** [comment]
I'm not a big fan of having three completely separate implementations for doing the same thing (`Bridge`/`BridgeLayer`, `LaysTerrain`/`BuildableTerrainLayer`, and now `ChangesTerrain`).

There's got to be a better way forward than this...

File: OpenRA.Mods.Common/Traits/ChangesTerrain.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	class ChangesTerrain : INotifyAddedToWorld
+	{
+		readonly ChangesTerrainInfo info;
+
+		public ChangesTerrain(ChangesTerrainInfo info)
+		{
+			this.info = info;
+		}
+
+		void INotifyAddedToWorld.AddedToWorld(Actor self)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/ChangesTerrain.cs]: You will need to clean up after yourself on removal, and think about what happens if CustomTerrain is already non-null.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/ChangesTerrain.cs]: Not for indestructible railways at the moment, but yes indeed later on that will be required.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/ChangesTerrain.cs]: This is a general purpose trait that will be available to all modders, so i'm not really happy with the idea of merging this with bugs.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/ChangesTerrain.cs]: You are right. This is an unnecessary limitation. Fixed.

**Comment by abcdefg30** [comment]
What about mobile units?
Also https://github.com/OpenRA/OpenRA/pull/11774#r73786035:
> [...] and think about what happens if CustomTerrain is already non-null.

**Comment by pchote** [comment]
> What about mobile units?

The `Requires<ImmobileInfo>` addresses that.

File: OpenRA.Mods.Common/Traits/ChangesTerrain.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+			var map = self.World.Map;
+			var terrain = map.Rules.TileSet.GetTerrainIndex(info.TerrainType);
+			previousTerrain = map.CustomTerrain[cell];
+			map.CustomTerrain[cell] = terrain;
+		}
+
+		void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
+		{
+			var cell = self.Location;
+			var map = self.World.Map;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/ChangesTerrain.cs]: These are redundant, just use them inline.

**Comment by LipkeGu** [comment]
>I'm not a big fan of having three completely separate implementations for doing the same thing >(Bridge/LowBridge/BridgeLayer, LaysTerrain/BuildableTerrainLayer, and now ChangesTerrain).
>
>There's got to be a better way forward than this...

ChangeTerrainLayer for all would be good for all three

File: OpenRA.Mods.Common/Traits/ChangesTerrain.cs (hunk 2 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/ChangesTerrain.cs]: They make the code easier to read for me.

**Review by abcdefg30 — APPROVED** [review]

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/212324080959dc53a9f78990694911ee7e7ef73b

**Pull request merged**